### ISSUE TITLE
Setting to go to end of (visual) line first

### ISSUE THREAD

**Issue by ralesk** [SublimeText/GoToEndOfLineOrScope#1]
First of all, thanks for making this plugin, it's amazing for getting a "smart End" feature in Sublime, especially in code with lots of line-end comments.

One of the things I like in the standard End key in Sublime though, when word wrap is on, is that it first goes to the end of the visual line, then to the hard EOL.

It would be really cool if this module could replicate that: go to visual eol, go to hard eol, go to eol-before-scope.

**Comment by keith-hall** [comment]
Hi, thanks for the feedback and the suggestion. I think it is a great idea and I will (hopefully within the next week or so) try to investigate adding support for visual eol.

**Comment by keith-hall** [comment]
Hi, I have just published a new release which will change the behavior of this plugin to default to going to visual eol, like the default <kbd>End</kbd> key binding which this plugin replaces, instead of hard eol. It may take a while for Package Control to pick up on it, but please try it out when you get chance and let me know how it goes.

(I didn't implement a setting to allow retaining the old behavior but I can do if anyone will want it back. AFAIK this plugin isn't very popular so I doubt anyone else will notice the change)

**Comment by ralesk** [comment]
It's kinda weird, it now _only_ goes to visual eol, and no further.
For example:
![image](https://user-images.githubusercontent.com/1564502/69425157-b3875600-0d2a-11ea-9fdb-6df3667bff68.png)
One End would go to after the `(` in `function(`, next End would go after the `{`
Currently, the plugin's End only goes to after the opening parenthesis.

Same thing happens with comments, it's impossible to go to the actual line end now.

By contrast, the built-in Home will visit every visual bol before the cursor position and then switch between code-bol and hard bol.

Comments behave a little differently I guess because the code-eol (comment start) can be both preceded by visual eols and followed by them.  I suppose standing anywhere in code should start visiting any visual eol after the cursor position, additionally stop at the code-eol (comment start), and continue stopping at visual eols, with pressing End on the hard-eol bouncing the cursor back to code-eol.

(Yeah, long lines and word wraps are evil, we don't prefer them either, but sometimes you just gotta juggle what you have :D)

**Comment by keith-hall** [comment]
good catch, thanks for the detailed description. Please accept my apologies for losing hard eol functionality and not testing thoroughly enough myself. I hope to get chance to rectify this in about 8 hours time - sorry for the inconvenience.

**Comment by ralesk** [comment]
No worries :D

**Comment by keith-hall** [comment]
> standing anywhere in code should start visiting any visual eol after the cursor position, additionally stop at the code-eol (comment start), and continue stopping at visual eols, with pressing End on the hard-eol bouncing the cursor back to code-eol.

I have made some improvements in f1e442d which mostly implements what you describe. So far I kept the default ST behavior of pressing end at visual eol jumping to hard eol, because I need a little more time to work on jumping to the next visual eol, which I don't have right now unfortunately.  I'm probably not much good at explaining, but hopefully the code is clearer now and you'll see from a little experimentation exactly what it is doing. Please let me know how it goes :) (though it may take a while for Package Control to see the updated version).

**Comment by ralesk** [comment]
Seems to work much better, thanks! :)
